### PR TITLE
Add fancy progress bar

### PR DESCRIPTION
**Pull request by LiamPattinson**
Since https://github.com/PlasmaFAIR/fortitude/pull/127, diagnostics are not printed as they are created, but instead are printed all at once after checking all files. On large projects, this can give the impression that the program is hanging, and gives no indication of when the program might be done.

This adds a fancy progress bar so the user can keep track of how things are going. Uses [indicatif](https://docs.rs/indicatif/latest/indicatif/), and the bar seems to nicely delete itself after completion.

### CODE REVIEW

**Comment by ZedThree** [comment]
Neat! The contrast is not great in my terminal though, I just get a blue bar appear briefly. Could we adjust that to be clearer?

**Comment by LiamPattinson** [comment]
Please could you check if this is better? Indicatif uses [console::Color](https://docs.rs/console/0.14.1/console/enum.Color.html) for styling, so at best you can provide an 8-bit colour (that's 8 bits for the whole colour, not 8 bits for each of red, green and blue!). I couldn't figure out what their 8-bit colour scheme was, but I came to this by trial and error.

**Comment by LiamPattinson** [comment]
Currently looks like this in my terminal:

Screenshot from 2024-11-22 15-33-38

One issue is that it vanishes too quickly for most projects. It would be nice if there were an option to only show it if it takes longer than 0.5s and the ETA is over 1s (or something like that).

**Comment by ZedThree** [comment]
Yep, that looks much better, thanks! Although the ETA doesn't work for me, just shows `[00:00]` even for large projects.

> One issue is that it vanishes too quickly for most projects. It would be nice if there were an option to only show it if it takes longer than 0.5s and the ETA is over 1s (or something like that).

Could only show it if we have more than, say, 100 files? We know that's only going to take less than a second.

Probably we should also have an option like `--no-progress-bar` to turn it off.

We might need to be a little careful about this when we get to the LSP server.

**Comment by LiamPattinson** [comment]
>  Although the ETA doesn't work for me, just shows [00:00] even for large projects.

I experimented with it a bit more and its usefulness seems to vary. On large projects and a single thread it's quite accurate, but on 4 threads it just stays on 0s, maybe bumping up to 1s. I've reverted it for now.

> Could only show it if we have more than, say, 100 files? We know that's only going to take less than a second.

I'm not sure where to draw that line, because there could be 50 gigantic files in one project that take an age to parse. Plus people could be running this on really slow hardware.

> Probably we should also have an option like `--no-progress-bar` to turn it off.

Done :+1: 

> We might need to be a little careful about this when we get to the LSP server.

Currently I've set it to make a progress bar with a style of `""` under some circumstances. It's not technically switching it off, so that might cause some problems down the line. I think that's something we can deal with later though.

**Review by ZedThree — APPROVED** [review]

File: fortitude/src/check.rs [hunk 1]
```diff
@@ -119,6 +120,7 @@ pub struct CheckSettings {
     pub line_length: usize,
     pub file_extensions: Vec<String>,
     pub output_format: OutputFormat,
+    pub no_progress_bar: bool,
```
Comment (ZedThree) [inline fortitude/src/check.rs:123]: If we want to be really rusty (rustic?), this should probably be a new enum:

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, CacheKey, is_macro::Is)]
pub enum ProgressBar {
    #[default]
    Disabled,
    Enabled,
}

impl From<bool> for ProgressBar {
    fn from(version: bool) -> Self {
        if version {
            ProgressBar::Enabled
        } else {
            ProgressBar::Disabled
        }
    }
}

impl Display for ProgressBar {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Disabled => write!(f, "disabled"),
            Self::Enabled => write!(f, "enabled"),
        }
    }
}
```

(just copy-pasting from `PreviewMode`)

This just gives type-safety if passing this around -- can't accidentally pass it to `PreviewMode` for example!

Comment (LiamPattinson) [reply fortitude/src/check.rs]: I've now done this, but went with three options: `Off`, `Ascii`, and `Fancy`.

**Comment by LiamPattinson** [comment]
A further update: Added ASCII mode, turned the bar off by default, and used alternative unicode characters to achieve half-character resolution (You can get even better if you go for a maximum height bar, but I think the narrow bar looks sleeker).

I've looked into it, and the bar doesn't get printed at all if you're not writing to terminal, so hopefully this shouldn't interfere with things like LSP.

**Review by ZedThree — APPROVED** [review]

**Pull request merged**